### PR TITLE
feat: Add generators in all the sage dimension sync

### PR DESCRIPTION
**Pull request by Ashutosh619-sudo**
https://app.clickup.com/1864988/v/l/6-901603904304-1

## Summary by CodeRabbit

- **New Features**
	- Introduced a method to retrieve the latest synchronization timestamps for various attributes, enhancing data accuracy.
	- Improved data retrieval methods for accounts, departments, expense types, and more, increasing efficiency by processing data in chunks.
	- Added a new test function to validate the creation and management of destination attributes.

- **Bug Fixes**
	- Adjusted test cases to align with updated data structures, ensuring accurate data handling in tests.

- **Documentation**
	- Updated return structures in test fixtures to reflect changes in data organization, requiring downstream adjustments.

### CODE REVIEW

**Review by ruuushhh — APPROVED** [review]

File: apps/sage_intacct/utils.py [hunk 1]
```diff
@@ -96,7 +96,9 @@ def sync_accounts(self):
         """
         Get accounts
         """
-        accounts = self.connection.accounts.get_all()
+
+        latest_sync = DestinationAttribute.objects.filter(workspace_id=self.workspace_id, attribute_type= 'ACCOUNT').order_by('-updated_at').first()
```
Comment (ashwin1111) [inline apps/sage_intacct/utils.py:100]: move this to a separate func - repeated in all sync func

File: apps/sage_intacct/utils.py [hunk 2]
```diff
@@ -96,7 +96,9 @@ def sync_accounts(self):
...
+        account_generator = self.connection.accounts.get_all_generator(updated_at=latest_sync.updated_at if latest_sync else None)
```
Comment (ashwin1111) [inline apps/sage_intacct/utils.py:101]: let's also send send status filter, it should be active. Need not check in code in 119 line

File: apps/sage_intacct/utils.py [hunk 3]
```diff
@@ -174,7 +179,10 @@ def sync_expense_types(self):
         """
         Get expense types
         """
-        expense_types = self.connection.expense_types.get_all()
+
+        latest_sync = DestinationAttribute.objects.filter(workspace_id=self.workspace_id, attribute_type='EXPENSE_TYPE').order_by('-updated_at').first()
+
+        expense_type_generator = self.connection.expense_types.get_all_generator(updated_at=latest_sync.updated_at if latest_sync else None)
```
Comment (ashwin1111) [inline apps/sage_intacct/utils.py:185]: status filter

File: apps/sage_intacct/utils.py (hunk 2 above)
Comment (ashwin1111) [reply apps/sage_intacct/utils.py:101]: let's also do `latest_sync.updated_at - 1 hour` to be safe. Also check if timezone could be a problem.

File: apps/sage_intacct/utils.py [hunk 4]
```diff
@@ -297,7 +308,8 @@ def sync_projects(self):
         """
         projects_count = self.connection.projects.count()
         if projects_count < SYNC_UPPER_LIMIT['projects']:
-            projects = self.connection.projects.get_all()
+            latest_sync = DestinationAttribute.objects.filter(workspace_id=self.workspace_id, attribute_type='PROJECT').order_by('-updated_at').first()
+            project_generator = self.connection.projects.get_all_generator(updated_at=latest_sync.updated_at if latest_sync else None)
```
Comment (ashwin1111) [inline apps/sage_intacct/utils.py:312]: status filter

File: apps/sage_intacct/utils.py [hunk 5]
```diff
@@ -449,21 +457,22 @@ def sync_expense_payment_types(self):
...
-            })
+        for expense_payment_types in expense_payment_type_generator:
+            for expense_payment_type in expense_payment_types:
+                expense_payment_type_attributes.append({
+                    'attribute_type': 'EXPENSE_PAYMENT_TYPE',
+                    'display_name': 'expense payment type',
+                    'value': expense_payment_type['NAME'],
+                    'destination_id': expense_payment_type['RECORDNO'],
+                    'detail': {
+                        'is_reimbursable': True if expense_payment_type['NONREIMBURSABLE'] == 'false' else False
```
[the remark this answers is not in the document]
Comment (ashwin1111) [reply apps/sage_intacct/utils.py:484]: +1 @Ashutosh619-sudo

File: apps/sage_intacct/utils.py (hunk 2 above)
Comment (ashwin1111) [reply apps/sage_intacct/utils.py:101]: let's also send `fields` value to generator to get only required fields that we use currently

Comment (ashwin1111) [reply apps/sage_intacct/utils.py:101]: these comments apply for every sync methods

File: apps/sage_intacct/utils.py [hunk 6]
```diff
@@ -668,7 +698,9 @@ def get_or_create_employee(self, source_employee: ExpenseAttribute):
             employee = None
 
         if not employee:
+            print(source_employee)
             created_employee = self.post_employees(source_employee)
+            print(created_employee)
```
Comment (ashwin1111) [inline apps/sage_intacct/utils.py:703]: remove loggers

File: apps/sage_intacct/utils.py [hunk 7]
```diff
@@ -799,7 +831,7 @@ def sync_vendors(self):
         """
         Get vendors
         """
-        vendors = self.connection.vendors.get_all()
+        vendor_generator = self.connection.vendors.get_all_generator()
```
Comment (ashwin1111) [inline apps/sage_intacct/utils.py:834]: updated at, fields, status to be added

File: apps/sage_intacct/utils.py [hunk 8]
```diff
@@ -92,13 +92,24 @@ def get_tax_exclusive_amount(self, amount, default_tax_code_id):
 
         return tax_exclusive_amount, tax_amount
 
+
+    def get_latest_sync(self, workspace_id, attribute_type):
+
+        latest_sync = DestinationAttribute.objects.filter(workspace_id=workspace_id, attribute_type=attribute_type).order_by('-updated_at').first()
+        if latest_sync:
+            return latest_sync.updated_at
```
Comment (ashwin1111) [inline apps/sage_intacct/utils.py]: let's do latest_sync.updated_at - 1 hour

File: apps/sage_intacct/utils.py [hunk 9]
```diff
@@ -92,13 +92,24 @@ def get_tax_exclusive_amount(self, amount, default_tax_code_id):
...
     def sync_accounts(self):
         """
         Get accounts
         """
 
-        latest_sync = DestinationAttribute.objects.filter(workspace_id=self.workspace_id, attribute_type= 'ACCOUNT').order_by('-updated_at').first()
-        account_generator = self.connection.accounts.get_all_generator(updated_at=latest_sync.updated_at if latest_sync else None)
+        fields = ['TITLE', 'ACCOUNTNO', 'ACCOUNTTYPE']
+        latest_updated_at= self.get_latest_sync(workspace_id=self.workspace_id, attribute_type='ACCOUNT')
+        account_generator = self.connection.accounts.get_all_generator(field='STATUS', value='active', fields=fields, updated_at=latest_updated_at-timedelta(hours=1) if latest_updated_at else None)
```
Comment (ashwin1111) [inline apps/sage_intacct/utils.py]: do -1hr in parent func instead of doing it everywhere

File: apps/sage_intacct/utils.py [hunk 10]
```diff
@@ -207,7 +206,9 @@ def sync_expense_types(self):
                 'detail': destination_attribute['detail']
             }
 
+        print('syncing expense types')
```
Comment (ashwin1111) [inline apps/sage_intacct/utils.py]: remove logger

File: apps/sage_intacct/utils.py [hunk 11]
```diff
@@ -207,7 +206,9 @@ def sync_expense_types(self):
...
         for expense_types in expense_type_generator:
+            print(expense_types)
```
Comment (ashwin1111) [inline apps/sage_intacct/utils.py]: remove logger

File: apps/sage_intacct/utils.py [hunk 12]
```diff
@@ -114,8 +115,10 @@ def sync_accounts(self):
         """
 
         fields = ['TITLE', 'ACCOUNTNO', 'ACCOUNTTYPE']
-        latest_updated_at= self.get_latest_sync(workspace_id=self.workspace_id, attribute_type='ACCOUNT')
-        account_generator = self.connection.accounts.get_all_generator(field='STATUS', value='active', fields=fields, updated_at=latest_updated_at-timedelta(hours=1) if latest_updated_at else None)
+        latest_updated_at = self.get_latest_sync(workspace_id=self.workspace_id, attribute_type='ACCOUNT')
+        print('latest_updated_at', latest_updated_at)
+        print('updated latest_updated_at',latest_updated_at)
```
Comment (ashwin1111) [inline apps/sage_intacct/utils.py]: remove logger

File: apps/sage_intacct/utils.py [hunk 13]
```diff
@@ -592,8 +530,14 @@ def sync_allocations(self):
...
         allocation_attributes = []
-        latest_sync = DestinationAttribute.objects.filter(workspace_id=self.workspace_id, attribute_type='ALLOCATION').order_by('-updated_at').first()
-        allocations_generator = self.connection.allocations.get_all_generator(field='STATUS', value='active', updated_at=latest_sync.updated_at if latest_sync else None)
+        latest_updated_at = self.get_latest_sync(workspace_id=self.workspace_id, attribute_type='ALLOCATION')
+        params = {}
+        if latest_updated_at:
+            params['updated_at'] = latest_updated_at
+        else:
+            params['field'] = 'STATUS'
+            params['value'] = 'active'
```
Comment (ashwin1111) [inline apps/sage_intacct/utils.py:539]: move params generation to a common func - it's same everywhere

File: apps/sage_intacct/utils.py [hunk 14]
```diff
@@ -358,55 +317,34 @@ def sync_projects(self):
...
-                        project_attributes.append({
-                            'attribute_type': 'PROJECT',
-                            'display_name': 'project',
-                            'value': project['NAME'],
-                            'destination_id': project['PROJECTID'],
-                            'active': True,
-                            'detail': detail
-                        })
-                        if project['PROJECTID'] in disabled_fields_map:
-                            disabled_fields_map.pop(project['PROJECTID'])
```
Comment (ashwin1111) [inline apps/sage_intacct/utils.py]: this has to be done for vendors sync too

File: apps/sage_intacct/utils.py (hunk 7 above)
Comment (ashwin1111) [reply apps/sage_intacct/utils.py:838]: missing fields, field, value, updated_at, etc

File: apps/sage_intacct/utils.py [hunk 15]
```diff
@@ -857,22 +848,23 @@ def sync_vendors(self):
...
+                        'value': vendor['NAME'],
+                        'destination_id': vendor['VENDORID'],
+                        'detail': detail,
+                        'active': True
+                    })
 
-                if vendor['VENDORID'] in disabled_fields_map:
-                    disabled_fields_map.pop(vendor['VENDORID'])
+                    if vendor['VENDORID'] in disabled_fields_map:
+                        disabled_fields_map.pop(vendor['VENDORID'])
```
Comment (ashwin1111) [inline apps/sage_intacct/utils.py:867]: has to be changed to new way

File: apps/sage_intacct/utils.py [hunk 16]
```diff
@@ -109,6 +109,17 @@ def get_latest_sync(self, workspace_id, attribute_type):
         return None
 
 
+    def generate_params(self, fields, latest_updated_at):
```
Comment (ashwin1111) [inline apps/sage_intacct/utils.py:112]: better name, construct_get_all_generator_params()

File: apps/sage_intacct/utils.py [hunk 17]
```diff
@@ -871,19 +851,6 @@ def sync_vendors(self):
                     'active': True
```
Comment (ashwin1111) [inline apps/sage_intacct/utils.py:851]: status=='active' needs to be added

**Review by ashwin1111 — APPROVED** [review]

File: apps/sage_intacct/utils.py (hunk 17 above)
Comment (Ashutosh619-sudo) [reply apps/sage_intacct/utils.py:851]: yup missed adding

**Pull request merged**